msg-socket/src/sub/driver.rs

```diff
@@ -155,124 +182,93 @@ where
+
         // Get the shared session stats.
         let session_stats = publisher_session.stats();
 
+        // Spawn the publisher session
+        tokio::spawn(publisher_session);
+
         for topic in self.subscribed_topics.iter() {
-            publisher_session.subscribe(topic.clone());
+            if publisher_channel
+                .try_send(SessionCommand::Subscribe(topic.clone()))
+                .is_err()
+            {
+                error!(publisher = %addr, "Error trying to subscribe to topic {topic} on startup: publisher channel closed / full");
+            }
         }
 
-        self.publishers.insert(addr, publisher_session);
+        self.publishers.insert(addr, publisher_channel);
         self.state.stats.insert(addr, session_stats);
     }
-}
-
-/// Manages the state of a single publisher, represented as a [`Stream`].
-#[must_use = "streams do nothing unless polled"]
-pub(super) struct PublisherSession<Io> {
-    /// The addr of the publisher
-    addr: SocketAddr,
-    /// The egress queue (for subscribe / unsubscribe messages)
-    egress: VecDeque<pubsub::Message>,
-    /// The inner stream
-    stream: PublisherStream<Io>,
-    /// The session stats
-    stats: Arc<SessionStats>,
-}
-
-impl<Io: AsyncRead + AsyncWrite + Send + Unpin> PublisherSession<Io> {
-    fn new(addr: SocketAddr, stream: PublisherStream<Io>) -> Self {
-        Self {
-            addr,
-            stream,
-            egress: VecDeque::with_capacity(4),
-            stats: Arc::new(SessionStats::default()),
-        }
-    }
-
-    /// Returns a reference to the session stats.
-    fn stats(&self) -> Arc<SessionStats> {
-        Arc::clone(&self.stats)
-    }
-
-    /// Queues a subscribe message for this publisher.
-    /// On the next poll, the message will be attempted to be sent.
-    fn subscribe(&mut self, topic: String) {
-        self.egress
-            .push_back(pubsub::Message::new_sub(Bytes::from(topic)));
-    }
-
-    /// Queues an unsubscribe message for this publisher.
-    /// On the next poll, the message will be attempted to be sent.
-    fn unsubscribe(&mut self, topic: String) {
-        self.egress
-            .push_back(pubsub::Message::new_unsub(Bytes::from(topic)));
-    }
-}
 
-impl<Io: AsyncRead + AsyncWrite + Unpin> Stream for PublisherSession<Io> {
-    type Item = Result<TopicMessage, pubsub::Error>;
+    /// Polls all the publisher channels for new messages. On new messages, forwards them to the socket.
+    /// If a publisher channel is closed, it will be removed from the list of publishers.
+    ///
+    /// Returns `Poll::Ready` if any progress was made and this method should be called again.
+    /// Returns `Poll::Pending` if no progress was made.
+    fn poll_publishers(&mut self, cx: &mut Context<'_>) -> Poll<()> {
+        let mut progress = false;
+
+        let mut inactive = Vec::new();
+
+        for (addr, rx) in self.publishers.iter_mut() {
```

Comment:
If I understand correctly, before we were polling a StreamMap with `poll_next_unpin` and now we are first looping through each publisher and polling each. Is this right?

Reply:
We're not polling the publisher directly, only the `Channel` that contains messages from the publisher. The publisher runs in a separate task and will be polled by the tokio scheduler